### DISCUSSION TITLE
downloading-artifacts-from-azure-devops-using-dotnet/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#49] [Announcements]
# downloading-artifacts-from-azure-devops-using-dotnet/

In this post I show how to use the Azure DevOps REST API to view the results of builds from Pipelines and how to download the artifacts

https://andrewlock.net/downloading-artifacts-from-azure-devops-using-dotnet/

**Comment by owaiskhanafridi** [comment]
The examples source code couldn't be found on mentioned link (https://github.com/DataDog/dd-trace-dotnet/blob/master/build/_build/Build.GitHub.cs#L261)

 I just want to give a url, artifact name, and download the artifacts through .NET code. I was trying to execute few powershell commands since artifacts can be downloaded from that but through this method the .NET app can directly communicate with the Azure Devops Artifacts.

**Comment by j3270** [comment]
This stackoverflow post appears to point to why 'GetArtifactContentZipAsync' doesn't work, at least lines up with what I am observing.  In your pipeline, if you publish artifacts as pipeline artifacts, which is the suggested way now to my understanding, 'GetArtifactContentZipAsync'  doesn't work.  Artifacts have to be published as build artifacts for 'GetArtifactContentZipAsync' to work, apparently.  I didn't change up my yaml to test this, but it makes sense to me.

https://stackoverflow.com/questions/62226816/vsts-buildclient-getartifactcontentzipasync-throws-microsoft-visualstudio-servic

I am trying out your work around and I get an empty zip file.  I will put up what I have in a bit.  Would appreciate any feedback you may have.

**Reply by j3270** [reply]
My attempt to download zip results in an empty zip file.  The artifact exists and is valid, I can download through Azure Devops platform UI.  The artifact itself is a zip file, is that a problem?  

Basically the same as your example with a few differences due to me using older framework.  ANy thoughts, comments on why my zip file is empty?

   '''

        static private async Task DownloadArtifactWithHttpClient(string artifactDownloadUrl, string artifactName, string downloadPath)
        {
            try
            {
                var localZipFilePath = downloadPath + $"\\{artifactName}.zip";

                // Use HttpClient to download the artifact
                using (var tempClient = new HttpClient())
                {
                    // Send request to the DownloadUrl specificed in the BuildArtifact
                    HttpResponseMessage response = await tempClient.GetAsync(artifactDownloadUrl);

                    if (!response.IsSuccessStatusCode)
                    {
                        // Something went wrong, shouldn't happen
                        throw new Exception($"Error downloading artifact: {response.StatusCode}:{response.ReasonPhrase}");
                    }

                    // Save the stream to a file
                    using (Stream zipFile = File.Create(localZipFilePath))
                    {
                        await response.Content.CopyToAsync(zipFile);
                    }
                }

                // All done!
                Console.WriteLine($"Done downloading too, {localZipFilePath}");
            }
            catch (Exception ex)
            {
                Console.WriteLine("Exception: " + ex.Message);
                if (ex.InnerException != null) Console.WriteLine("Detailed Info: " + ex.InnerException.Message);
                Console.WriteLine("Stack:\n" + ex.StackTrace);
            }
            
        }
   '''

**Reply by j3270** [reply]
I realized my mistake.  I wasn't authenticating the new HttpClient.

Earlier in the example we authenticate the 'Microsoft.VisualStudio.Services.WebApi.VssConnection' with our PAT and get a BuildHttpClient to work with.  That client was authenticated and what we used to get the download URL.

The new HttpClient isn't authenticated in the example.  However,  I kept digging and looked at the REST API offered by MS and they show how to use the HttpClient and how to authenticate it with a PAT.  After doing that in my function above, my downloads succeeded.

I should have realized this from the 203 response code, but I am learning.

Thank you for your article, it got me 90% of the way there.

[a comment by an automated account is not shown]

**Comment by TheFrogPrince** [comment]
So... on the subject of actually downloading the artifact, the above code does not work as is.  As pointed out in other threads, you need to separately authenticate your HttpClient, which I find to be annoying when I am using the interactive authentication method.  

Never fear!

This is the part of .NET that I truly love.  Under the hood, the BuildHttpClient already has it's own copy of the HttpClient, and we just need to use that one. (the code below is not complete, but gives you the idea about how to use reflection to get the HttpClient that is already wired up):

            BuildHttpClient oClient = oConn.GetClient<BuildHttpClient>();

            Build oBuild = await oClient.GetBuildAsync(oProject.Id, 1234);

            // here we use reflection to get the HttpClient that is already wired up and authenticated
            System.Reflection.PropertyInfo oClientInfo = oClient.GetType().GetProperty("Client", BindingFlags.Instance | BindingFlags.NonPublic | BindingFlags.FlattenHierarchy);
            HttpClient oWebClient = (HttpClient)oClientInfo.GetValue(oClient);

            List<BuildArtifact> oArtifacts = await oClient.GetArtifactsAsync(oProject.Id, oBuild.Id);
            BuildArtifact oArtifact = oArtifacts[0];

            using (Stream oDownloadStream = await oWebClient.GetStreamAsync(oArtifact.Resource.DownloadUrl))
                using( ZipArchive oZip = new ZipArchive(oDownloadStream))
                {
                    foreach (ZipArchiveEntry oEntry in oZip.Entries)
                    {
                        if (oEntry.Length > 0)
                        {
                            using (Stream oInStream = oEntry.Open())
                            {
                                StreamReader oReader = new StreamReader(oInStream);
                                string sJson = await oReader.ReadToEndAsync();

                                Models.Root oResults = Newtonsoft.Json.JsonConvert.DeserializeObject<Models.Root>(sJson);
                                foreach( Models.Run oRun in oResults.runs)
                                {
                                    if (oRun.results.Count > 0)
                                    {

                                    }
                                }
                            }
                        }
                    }
                }
            }
